pkg/remediator/watch/listwatch.go

```diff
@@ -0,0 +1,299 @@
+		case meta.RESTScopeNameNamespace:
+			if namespace == "" {
+				// all namespaces
+				return dynamicClient.Resource(mapping.Resource).Watch(ctx, options)
+			}
+			return dynamicClient.Resource(mapping.Resource).Namespace(namespace).Watch(ctx, options)
+		default:
+			return nil, fmt.Errorf("invalid resource scope %q for resource %q", mapping.Scope, mapping.Resource)
+		}
+	}
+	return &ListWatch{ListFunc: listFunc, WatchFunc: watchFunc}
+}
+
+// List a set of apiserver resources
+func (lw *ListWatch) List(ctx context.Context, options metav1.ListOptions) (*unstructured.UnstructuredList, error) {
+	// ListWatch is used in Reflector, which already supports pagination.
+	// Don't paginate here to avoid duplication.
+	return lw.ListFunc(ctx, options)
+}
+
+// Watch a set of apiserver resources
+func (lw *ListWatch) Watch(ctx context.Context, options metav1.ListOptions) (watch.Interface, error) {
+	return lw.WatchFunc(ctx, options)
+}
+
+type listAndWatch struct {
+	stopOnce sync.Once
+	stopCh   chan struct{}
+	eventCh  chan watch.Event
+}
+
+func (ww *listAndWatch) Stop() {
+	ww.stopOnce.Do(func() {
+		close(ww.stopCh)
+	})
+	<-ww.eventCh
+}
+
+func (ww *listAndWatch) ResultChan() <-chan watch.Event {
+	return ww.eventCh
+}
+
+// ListAndWatch wraps a list and watch and returns a watch interface.
+// This way, you can watch from "now" and get all the existing objects, not just
+// new changes to objects.
+//
+// If ResourceVersion is specified, ListAndWatch acts like a normal Watch.
+// Otherwise, a List is performed first and `Added` events are simulated,
+// followed by a Watch, where subsequent `Added` events for pre-existing objects
+// are converted to `Modified` events, unless that object is deleted first.
+func ListAndWatch(ctx context.Context, lw ListerWatcher, opts metav1.ListOptions) (watch.Interface, error) {
+	if opts.ResourceVersion != "" {
+		// Just watch!
+		return lw.Watch(ctx, opts)
+	}
+
+	opts.AllowWatchBookmarks = false // not allowed for List
+	opts.Watch = false               // not allowed for List
+
+	// TODO: Use a short timeout for List and longer for Watch
+
+	cObjList, err := lw.List(ctx, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	// Specify ResourceVersion to ensure the Watch starts where the List stopped.
+	// watch.Until will update the ListOptions for us.
+	opts.ResourceVersion = cObjList.GetResourceVersion()
+
+	// Enable bookmarks to ensure retries start from the latest ResourceVersion,
+	// even if there haven't been any updates to the object being watched.
+	opts.AllowWatchBookmarks = true
+	opts.Watch = true
+
+	wrapper := &listAndWatch{
+		stopCh:  make(chan struct{}),
+		eventCh: make(chan watch.Event),
+	}
+
+	// Spawn background thread to handle watching and pre-processing events.
+	go func() {
+		defer close(wrapper.eventCh)
+
+		// De-dupe Added events to Modified events by caching the IDs of pre-existing objects.
+		ids := make(map[core.ID]struct{}, len(cObjList.Items))
```

Comment:
Is this the primary motivation for list and watch? Namely differentiating pre-existing objects as `modified` rather than `added` for the watch?

Reply:
The motivation for doing a List and a Watch, instead of just a Watch, is to pull the current state so we don't effectively ignore changes before we started watching until the next event after we started watching.

This ListAndWatch is a convenience method that returns the same watch.Interface that Watch does, but it does the List first and translates the output into Added events. Then, to avoid duplicate Added events, we have to cache which ones we've seen already and convert new Added events into Modified events. And we also have to remove IDs from the cache if they're Deleted, so that we don't translate any subsequent Added event when it's re-created.

Reply:
The same ListAndWatch pattern can be seen in the Informer: https://github.com/kubernetes/client-go/blob/release-1.26/tools/cache/reflector.go#L256

The difference is that this one is standalone, uses Unstructured objects (dynamic.Client), cancels with context instead of channel, doesn't handle retries/resync,  and handles errors better (doesn't just log and retry).

If I wanted to go further, I'd write a DynamicInformer, but the FilteredWatcher code already handles retries. We could also later write something like the Until watch tools, but we don't need them yet.

Reply:
Actually, because this returns a watch.Interface, you can already use it with UntilWithoutRetry, if you want: https://github.com/kubernetes/client-go/blob/release-1.26/tools/watch/until.go#L56